### PR TITLE
Move conditional `useInsertionEffect` declarations into separate package

### PR DESCRIPTION
**Pull request by Andarist**
**Why?**

Sometimes people bundle Emotion in their libraries~ (for whatever reasons, one such example would be `@storybook/theming`) and while doing so their bundlers/minifiers replace our carefully crafted `React['useInsertion' + 'Effect']` with just `React.useInsertionEffect`. That, in turn, can create problems when consuming such bundles with webpack and some other tools (it's also the reason why we are using this weird pattern in the first place). Most notably, webpack errors with:
> require function is used in a way in which dependencies cannot be statically extracted

It's hard to control bundlers/minifiers to leave this pattern alone, so I think that what we can offer to people with such setups is this package that they will be able to externalize in their setups. The main reason for bundling Emotion etc is to isolate its context (create its copy) to avoid conflicts with 3rd party code and stuff like that - and it's not to actually make a copy of everything.

fixes https://github.com/emotion-js/emotion/issues/2738

### CODE REVIEW

File: packages/use-insertion-effect-with-fallbacks/src/index.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+const syncFallback = create => create()
+
+const useInsertionEffect = React['useInsertion' + 'Effect']
+  ? React['useInsertion' + 'Effect']
+  : false
+
+export const useInsertionEffectAlwaysWithSyncFallback = !isBrowser
+  ? syncFallback
+  : useInsertionEffect || syncFallback
```
Comment (Andarist) [inline packages/use-insertion-effect-with-fallbacks/src/index.js:13]: Please be extra careful when reviewing this as the shape of this has slightly changed when I moved this. I think that the current code is functionally equivalent to the old version though.

File: packages/use-insertion-effect-with-fallbacks/package.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "@emotion/use-insertion-effect-with-fallbacks",
```
Comment (Andarist) [inline packages/use-insertion-effect-with-fallbacks/package.json:2]: Should this be `*-with-fallback` or `*-with-fallbacks`? :sweat_smile: #realproblems

**Comment by gyhaLabs** [comment]
I have this very similar issue in our project. We are using a common UI library (of our own) - which uses material UI with @emotion/react - which may be a very common combo. (See the installation doc for the material UI: https://mui.com/material-ui/getting-started/installation/).

When I am including this common/shared UI package in the main project, it won't do the build/deployment and the GitLab pipline fails with the error below.

image

This is a valid use-case, right? It is a similar issue with the above mentioned, right @Andarist?

Thanks.

**Comment by Andarist** [comment]
@gyhaLabs I think this is very much related. My strong suspicion is that you are **bundling** Emotion (effectively creating its copy) and that, together with your UI library, is later consumed together by your consumers.

File: packages/use-insertion-effect-with-fallbacks/package.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  "version": "1.0.0",
```
Comment (mitchellhamilton) [inline packages/use-insertion-effect-with-fallbacks/package.json:3]:
```suggestion
  "version": "0.0.0",
```

+ a changeset

**Review by mitchellhamilton — APPROVED** [review]

**Pull request merged**